src/main/java/net/minecraftforge/common/context/ContextKey.java

```diff
@@ -0,0 +1,126 @@
+package net.minecraftforge.common.context;
+
+import java.util.Optional;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Consumer;
+
+import org.jetbrains.annotations.Nullable;
+
+import com.google.common.collect.MapMaker;
+
+import net.minecraft.resources.ResourceLocation;
+
+/**
+ * A Context Key is a key used when passing arbitrary context to a method.<br>
+ * It retains the ability to cast the underlying context Object to the target type.
+ * 
+ * @param <T> The type of the target context object.
+ */
+public final class ContextKey<T>
```

Comment:
I think we should get rid of this whole thing and just wait for 1.20.2 to fix the integer index.

Reply:
I think some sort of declared context is definitely better than just having it be a different integer index. Especially for mod compat when it comes to some mods may want to semi conditionally tick things in their inventories so then they can pass their own context or type of context rather than requiring mods to implement their own API to ensure that random things aren't being ticked that shouldn't be ticked in that type of context.

Reply:
This new method is expanded in scope to cover ticking in any inventory, but perhaps the context key system is too generic.  Some of the parameters could be reverted to native method params, and this could be reduced to a simple context interface that provides no methods by default (instead of using ContextKey)

Though I think keyed-contexts may be useful in other places as well.

Reply:
> some mods may want to semi conditionally tick things in their inventories so then they can pass their own context or type of context rather than requiring mods to implement their own API to ensure that random things aren't being ticked that shouldn't be ticked in that type of context.

But then they're still depending on a third party API (the type of context)...

Reply:
The context thing as a whole is too generic, and will be misused. It is pure technical debt.

Curios can keep providing its own API, I don't see the issue with that. It would still have to provide a context type anyway, so no difference...